### PR TITLE
Add repo2uast and uast2prox

### PR DESCRIPTION
**Pull request by fineguy**
repo2uast – convert each repository into a .asdf file with all extracted UASTs and filenames.
uast2prox – convert UASTs into proximity matrices (compatible with id2vec_preprocess step)

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@fineguy You need to rebase onto the latest `develop`

File: ast2vec/model2/proxbase.py [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+from ast2vec.token_parser import TokenParser
+from ast2vec.uast import UASTModel
+from ast2vec.model2.base import Model2Base
+
+EDGE_TYPES = ["r", "t", "rt", "R", "T", "RT"]
+
+
+class ProxBase(Model2Base):
+    """
+    Contains common utilities for proximity matrix models.
```
Comment (vmarkovtsev) [inline ast2vec/model2/proxbase.py]: Can you please write a few sentences what is a proximity matrix, what is the algorithm and so on.

File: ast2vec/model2/proxbase.py [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+
+import numpy
+from scipy.sparse import coo_matrix, diags
+
+from ast2vec.coocc import Cooccurrences
+from ast2vec.token_parser import TokenParser
+from ast2vec.uast import UASTModel
+from ast2vec.model2.base import Model2Base
+
+EDGE_TYPES = ["r", "t", "rt", "R", "T", "RT"]
```
Comment (vmarkovtsev) [inline ast2vec/model2/proxbase.py]: How about writing docs which tell wtf these are

File: ast2vec/model2/prox.py [hunk 3]
```diff
@@ -0,0 +1,40 @@
+from typing import Dict, Tuple
```
Comment (vmarkovtsev) [inline ast2vec/model2/prox.py]: The whole file contains no docs

File: ast2vec/model2/proxbase.py (hunk 2 above)
Comment (fineguy) [inline ast2vec/model2/proxbase.py]: Is it better to refer user to the description in `__main__.py` or copy it here?

File: ast2vec/__main__.py [hunk 4]
```diff
@@ -164,6 +196,30 @@ def main():
...
+        "--edges", nargs="+", default=EDGE_TYPES, choices=EDGE_TYPES,
+        help="If not specified, then node-to-node adjacency is assumed. Suppose we have two "
+        "connected nodes A and B:\n"
+        "r - connect node roles with each other.\n"
+        "t - connect node tokens with each other.\n"
+        "rt - connect node tokens with node roles.\n"
+        "R - connect node A roles with node B roles.\n"
+        "T - connect node A tokens with node B tokens.\n"
+        "RT - connect node A roles(tokens) with node B tokens(roles).")
+    uast2prox_parser.add_argument(
```
Comment (EgorBu) [inline ast2vec/__main__.py]: is search depth equal to 1?

File: ast2vec/model2/proxbase.py (hunk 2 above)
Comment (vmarkovtsev) [inline ast2vec/model2/proxbase.py]: Copy-paste rules for docs

File: ast2vec/model2/prox.py (hunk 3 above)
Comment (EgorBu) [inline ast2vec/model2/prox.py]: +1

File: ast2vec/model2/proxbase.py (hunk 2 above)
Comment (EgorBu) [inline ast2vec/model2/proxbase.py]: I think it's better to put description here and main should refer to this file (so docs should be in one place)

File: ast2vec/model2/proxbase.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+        super(ProxBase, self).__init__(*args, **kwargs)
+        self.edges = set(edges)
+        self.roles = list()
+        self.tokens = list()
+        self.role2ind = dict()
+        self.token2ind = dict()
+        self.dok_matrix = defaultdict(int)
+        self._token_parser = TokenParser()
+
+    def convert_model(self, model):
```
Comment (EgorBu) [inline ast2vec/model2/proxbase.py]: docstring

File: ast2vec/repo2/uast.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+from ast2vec.uast import UASTModel
+
+
+class Repo2UASTModel(Repo2Base):
+    """
+    Extract UASTs from repository
+    """
+    MODEL_CLASS = UASTModel
+
+    def convert_uasts(self, file_uast_generator):
```
Comment (EgorBu) [inline ast2vec/repo2/uast.py]: docstrings

File: ast2vec/repo2/uast.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+            uasts.append(file_uast.response)
+            filenames.append(file_uast.filename)
+
+        return filenames, uasts
+
+
+class Repo2UASTModelTransformer(RepoTransformer):
+    WORKER_CLASS = Repo2UASTModel
+
+    def dependencies(self):
```
Comment (EgorBu) [inline ast2vec/repo2/uast.py]: ??

File: ast2vec/repo2/uast.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+        return []
+
+    def result_to_model_kwargs(self, result, url_or_path):
```
Comment (EgorBu) [inline ast2vec/repo2/uast.py]: docstrings

File: ast2vec/repo2/uast.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        filenames, uasts = result
+        if len(filenames) == 0:
+            raise ValueError('No need to store empty model.')
```
Comment (EgorBu) [inline ast2vec/repo2/uast.py]: quotes

File: ast2vec/__main__.py (hunk 4 above)
Comment (fineguy) [inline ast2vec/__main__.py]: Yep

File: ast2vec/repo2/uast.py (hunk 7 above)
Comment (fineguy) [inline ast2vec/repo2/uast.py]: https://github.com/src-d/ast2vec/blob/86f38c6c42f1b2028f76f32dd7af7e5739dcae33/ast2vec/repo2/base.py#L325

**Pull request merged**